File: algorithms/cpp/movingStonesUntilConsecutiveII/movingStonesUntilConsecutiveII.cpp
// Source: https://leetcode.com/problems/moving-stones-until-consecutive-ii/
// Author: Miao Zhang
// Date:   2021-04-06

class Solution {
public:
    vector<int> numMovesStonesII(vector<int>& stones) {
        sort(begin(stones), end(stones));
        int n = stones.size();
        int i = 0;
        int low = n;
        for (int j = 0; j < n; j++) {
            while (stones[j] - stones[i] + 1 > n) i++;
            int already_stone = j - i + 1;
            if (already_stone == n - 1 && stones[j] - stones[i] + 1 == n - 1) {
                low = min(low, 2);
            } else {
                low = min(low, n - already_stone);
            }
        }
        return {low, max(stones[n - 1] - stones[1] - n + 2, stones[n - 2] - stones[0] - n + 2)};
    }
};

File: algorithms/python/movingStonesUntilConsecutiveII/movingStonesUntilConsecutiveII.py
#! /usr/bin/env python3
# -*- coding: utf-8 -*-

# Source: https://leetcode.com/problems/moving-stones-until-consecutive-ii/
# Author: Miao Zhang
# Date:   2021-04-06

class Solution:
    def numMovesStonesII(self, stones: List[int]) -> List[int]:
        stones.sort()
        n = len(stones)
        i = 0
        low = n
        for j in range(n):
            while stones[j] - stones[i] + 1 > n:
                i += 1
            if j - i + 1 == n - 1 and stones[j] - stones[i] + 1 == n - 1:
                low = min(low, 2)
            else:
                low = min(low, n - (j - i + 1))
        return [low, max(stones[-1] - stones[1] - n + 2, stones[n-2] - stones[0] - n + 2)]
            
